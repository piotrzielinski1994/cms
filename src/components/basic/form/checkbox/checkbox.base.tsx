import { forwardRef, InputHTMLAttributes } from 'react';
import Form from '../root/form';

type InputProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'value' | 'type'> & {
  name: string;
  value?: string;
};

const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  return <input ref={ref} type="checkbox" {...props} />;
});

Input.displayName = 'CheckboxBase.Input';

const CheckboxBase = {
  Root: Form.Group,
  Label: Form.Label,
  Input,
  Error: Form.Error,
};

export default CheckboxBase;
