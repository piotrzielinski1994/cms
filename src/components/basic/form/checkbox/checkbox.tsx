import { cn } from '@/utils/tailwind';
import { Check } from 'lucide-react';
import { InputHTMLAttributes } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import Form from '../root/form';

type CheckboxProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'name' | 'type'> & {
  label: string;
  name: string;
  error?: string;
};

type CheckboxContainerProps<T extends FieldValues> = Omit<CheckboxProps, 'name'> & {
  control: Control<T>;
  name: Path<T>;
};

const checkboxClassNames = {
  wrapper: cn(
    'group',
    'has-[:disabled]:text-foreground/50',
    'grid grid-cols-[auto_1fr] gap-2',
    'cursor-pointer has-[:disabled]:cursor-not-allowed',
  ),
  checkbox: ({ isValid }: { isValid: boolean }) => {
    return cn(
      'w-[1lh] h-[1lh] p-[0.15lh]',
      'border border-solid border-current bg-input',
      'group-hover:group-has-[input:enabled]:border-foreground/90',
      'group-hover:group-has-[input:enabled]:ring-foreground/90',
      { '[&:not(:focus)]:text-red-500': !isValid },
      'ring-inset focus-within:ring-1 ring-current',
    );
  },
  icon: cn('hidden group-has-[:checked]:block', 'w-[0.7lh] h-[0.7lh] -ml-[0.03lh]'),
};

const Checkbox = ({ label, error, className, ...props }: CheckboxProps) => {
  return (
    <Form.Group>
      <Form.Label className={checkboxClassNames.wrapper}>
        <div className={cn(checkboxClassNames.checkbox({ isValid: !error }), className)}>
          <input {...props} type="checkbox" className="sr-only" />
          <Check className={checkboxClassNames.icon} />
        </div>
        <span>{label}</span>
      </Form.Label>
      <Form.Error>{error}</Form.Error>
    </Form.Group>
  );
};

const CheckboxContainer = <T extends FieldValues>(props: CheckboxContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const { field, fieldState } = useController({ control, name });
  return <Checkbox error={fieldState.error?.message} {...rest} {...field} />;
};

export { Checkbox, CheckboxContainer };
