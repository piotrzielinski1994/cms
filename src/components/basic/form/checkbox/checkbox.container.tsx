import { ComponentProps } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import { Checkbox } from './checkbox';

type CheckboxContainerProps<T extends FieldValues> = {
  control: Control<T>;
  name: Path<T>;
} & Omit<ComponentProps<typeof Checkbox>, 'name'>;

const CheckboxContainer = <T extends FieldValues>(props: CheckboxContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const { field, fieldState } = useController({ control, name });
  return <Checkbox error={fieldState.error?.message} {...rest} {...field} />;
};

export { CheckboxContainer };
