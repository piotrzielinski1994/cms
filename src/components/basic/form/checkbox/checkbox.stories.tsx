import { DEFAULT_VALUE, getFallback } from '@/config/storybook/utils';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { useState, type ComponentProps } from 'react';
import { Checkbox as CheckboxComponent } from './checkbox';

type Args = ComponentProps<typeof CheckboxComponent>;

const ControlledCheckbox = (props: Args) => {
  const [isChecked, setIsChecked] = useState(false);
  return (
    <CheckboxComponent
      {...props}
      onChange={(e) => setIsChecked(e.currentTarget.checked)}
      checked={isChecked}
    />
  );
};

const meta: Meta<ComponentProps<typeof ControlledCheckbox>> = {
  component: ControlledCheckbox,
  title: 'Components/Basic/Form/Checkbox',
  argTypes: {
    label: { control: 'text' },
    error: { control: 'text' },
    disabled: { control: 'boolean' },
  },
  args: {
    label: DEFAULT_VALUE,
    disabled: false,
  },
};

const Render = ({ label, ...args }: Args) => {
  const t = useTranslations('storybook.basic.form.checkbox');

  return (
    <div className="grid gap-2">
      <ControlledCheckbox
        {...args}
        label={getFallback(label, t('default.label'))}
        id="input1"
        name="input1"
      />
      <CheckboxComponent id="input2" name="input2" label={t('disabled.label')} disabled />
      <CheckboxComponent
        {...args}
        id="input3"
        name="input3"
        label={t('invalid.label')}
        error={t('invalid.error')}
      />
    </div>
  );
};

const Checkbox: StoryObj<typeof CheckboxComponent> = { render: Render };

export { Checkbox };
export default meta;
