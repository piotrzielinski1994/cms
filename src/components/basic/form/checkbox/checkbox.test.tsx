import { render } from '@testing-library/react';
import { ComponentProps, useState } from 'react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';
import Form from '../root/form';
import { Checkbox } from './checkbox';

describe('Checkbox', () => {
  const defaultProps = {
    name: 'name',
    id: 'id',
    label: 'Label',
    onChange: vi.fn(),
  } satisfies ComponentProps<typeof Checkbox>;

  const ControlledComponent = (props: ComponentProps<typeof Checkbox>) => {
    const [isChecked, setIsChecked] = useState<boolean>(props.checked ?? false);
    return (
      <Form.Group>
        <Form.Label htmlFor={props.id}>Label</Form.Label>
        <Checkbox
          {...props}
          checked={isChecked}
          onChange={(e) => setIsChecked(e.currentTarget.checked)}
        />
      </Form.Group>
    );
  };

  it('should have no accessibility violations', async () => {
    const { container } = render(<ControlledComponent {...defaultProps} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(<ControlledComponent {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
