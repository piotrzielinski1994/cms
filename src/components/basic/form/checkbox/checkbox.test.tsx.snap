// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > should match the snapshot 1`] = `
<div>
  <div
    class="grid gap-1 content-start grid-rows-[auto_1fr]"
  >
    <label
      class="self-end"
      for="id"
    >
      Label
    </label>
    <div
      class="grid gap-1 content-start grid-rows-[auto_1fr]"
    >
      <label
        class="self-end group has-[:disabled]:text-foreground/50 grid grid-cols-[auto_1fr] gap-2 cursor-pointer has-[:disabled]:cursor-not-allowed"
      >
        <div
          class="w-[1lh] h-[1lh] border border-solid border-current bg-input group-hover:group-has-[input:enabled]:border-foreground/90 group-hover:group-has-[input:enabled]:ring-foreground/90 ring-inset focus-within:ring-1 ring-current grid place-items-center"
        >
          <input
            class="sr-only"
            id="id"
            name="name"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="lucide lucide-check hidden group-has-[:checked]:block w-[0.7lh] h-[0.7lh]"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 6 9 17l-5-5"
            />
          </svg>
        </div>
        <span>
          Label
        </span>
      </label>
      <span
        class="min-h-[1em] text-sm text-red-500 leading-none"
        role="alert"
      />
    </div>
  </div>
</div>
`;
